packages/teleterm/src/main.ts

```diff
@@ -31,13 +31,20 @@ const mainProcess = MainProcess.create({
   fileStorage,
 });
 
-// node-pty is not yet context aware
-app.allowRendererProcessReuse = false;
+const sharedProcess = fork(
```

Comment:
Did you consider documenting the whole communication flow somewhere?

PTY process communication is already quite complex, now we're adding another layer (layers?) of indirection and on top of that there's bidirectional streaming. My head hurt after reviewing this for over an hour!

I was thinking of a sequence diagram, created using something like https://www.websequencediagrams.com/ so that it can be extended if needed. But it wouldn't necessarily have to be a diagram.

My main points when trying to fit the code in my head is that there are various events that we register listeners to. Some of those events are sent from the client to the server, some from the server to the client and then there are also events created my the PTY process itself.

It'd be good to have a piece of documentation which could help you grok this faster.

Reply:
Please take a look at the [diagram](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgVW4AAwVkCgpEb2N1bWVudCBUZXJtaW5hbC0-UFRZIFByb2Nlc3MgQ2xpZW50OiB3YW50cyBuZXcgUFRZCgAQEgAwBkhvc3QgU2VydmVyOiBjcmVhdGVQdHkARQcob3B0aW9ucykAOgUAIAstAGYNOgBgBgAtCgAeEgCBDRVwdHlJZCBvZiB0aGUgcACBOAdpcyByZXR1cm5lZACBDyZlc3RhYmxpc2ggZXhjaGFuZ2VFdmVudHMoKSBjaGFubmVsCm5vdGUgbGVmdCBvZiAAghkTCmMAgi8FIGhhcyBiZWVuAIICB2QsIApzbwAeEyBjYW4gYXR0YWNoIGV2ADQGbmRsZXJzIAp0bwCBSgUAdAcgKG9uRGF0YS9vbk9wZW4vb25FeGl0KQplbmQgbm90ZQCDAxUAg1ARAIImBQCCGQlvYmplY3QAg18oc3RhcnQAgwsHAIIqIgCCNQ0uADYGAIMiBSkgAIJABnJpZ2gAgj4JAIQwDQoAgnEOAIIZB2VzAIITDwCCGQgAhSkKAIF_JQCENRgAgVcKAIVXDACFSRMAgnsGAIRzKACBeg8AMA0AgwwjAGQJAIcFKGRpc3Bvc2UAgn8pbmQAhV8PAIVjCQCCDx4AYQcAhmUJYW5kIHJlbW92ZSBpdAo&s=default).

Do you think it helps with understanding the flow?